### ISSUE TITLE
SNOW-856979: cryptography dependency version. security issue.

### ISSUE THREAD

**Issue by SiftRonnie** [snowflakedb/snowflake-connector-python#1629]
### Python version

3.6.9

### Operating system and processor architecture

Linux-4.15.0-123-generic-x86_64-with-Ubuntu-18.04-bionic

### Installed packages

```python
agate==1.7.0                                                                                                                                                                                                                                                                            asn1crypto==1.5.1                                                                                                                                                                                                                                                                       attrs==22.2.0                                                                                                                                                                                                                                                                           awswrangler==2.13.0                                                                                                                                                                                                                                                                     Babel==2.12.1                                                                                                                                                                                                                                                                           beautifulsoup4==4.12.0                                                                                                                                                                                                                                                                  betterproto==1.2.5                                                                                                                                                                                                                                                                      boto3==1.26.80                                                                                                                                                                                                                                                                          botocore==1.29.80                                                                                                                                                                                                                                                                       certifi==2022.12.7                                                                                                                                                                                                                                                                      cffi==1.15.1                                                                                                                                                                                                                                                                            charset-normalizer==3.1.0                                                                                                                                                                                                                                                               ciso8601==2.3.0                                                                                                                                                                                                                                                                         click==8.1.3                                                                                                                                                                                                                                                                            cloudpickle==2.2.1                                                                                                                                                                                                                                                                      cmdstanpy==1.1.0                                                                                                                                                                                                                                                                        colorama==0.4.6                                                                                                                                                                                                                                                                         contourpy==1.0.7                                                                                                                                                                                                                                                                        convertdate==2.4.0                                                                                                                                                                                                                                                                      coverage==7.2.2                                                                                                                                                                                                                                                                         cryptography==40.0.2                                                                                                                                                                                                                                                                    cycler==0.11.0                                                                                                                                                                                                                                                                          Cython==0.29.34                                                                                                                                                                                                                                                                         dask==2023.3.0                                                                                                                                                                                                                                                                          dbt-athena-community==1.4.1                                                                                                                                                                                                                                                             dbt-core==1.4.5                                                                                                                                                                                                                                                                         dbt-extractor==0.4.1                                                                                                                                                                                                                                                                    dbt-snowflake==1.4.3                                                                                                                                                                                                                                                                    decorator==5.1.1                                                                                                                                                                                                                                                                        deepdiff==5.8.1                                                                                                                                                                                                                                                                         ephem==4.1.4                                                                                                                                                                                                                                                                            et-xmlfile==1.1.0                                                                                                                                                                                                                                                                       exceptiongroup==1.1.1                                                                                                                                                                                                                                                                   filelock==3.12.2                                                                                                                                                                                                                                                                        fonttools==4.39.3                                                                                                                                                                                                                                                                       fsspec==2023.3.0                                                                                                                                                                                                                                                                        future==0.18.3                                                                                                                                                                                                                                                                          future-fstrings==1.2.0                                                                                                                                                                                                                                                                  grpclib==0.4.3                                                                                                                                                                                                                                                                          h2==4.1.0                                                                                                                                                                                                                                                                               h2o==3.34.0.3                                                                                                                                                                                                                                                                           hijri-converter==2.3.1                                                                                                                                                                                                                                                                  holidays==0.23                                                                                                                                                                                                                                                                          hologram==0.0.15                                                                                                                                                                                                                                                                        hpack==4.0.0                                                                                                                                                                                                                                                                            hyperframe==6.0.1                                                                                                                                                                                                                                                                       idna==3.4                                                                                                                                                                                                                                                                               importlib-metadata==6.7.0                                                                                                                                                                                                                                                               importlib-resources==5.12.0                                                                                                                                                                                                                                                             iniconfig==2.0.0                                                                                                                                                                                                                                                                        isodate==0.6.1                                                                                                                                                                                                                                                                          jaraco.classes==3.2.3                                                                                                                                                                                                                                                                   jeepney==0.8.0                                                                                                                                                                                                                                                                          Jinja2==3.1.2                                                                                                                                                                                                                                                                           jmespath==1.0.1                                                                                                                                                                                                                                                                         joblib==1.2.0                                                                                                                                                                                                                                                                           jsonpath-ng==1.5.3                                                                                                                                                                                                                                                                      jsonschema==3.2.0                                                                                                                                                                                                                                                                       keyring==23.13.1                                                                                                                                                                                                                                                                        kiwisolver==1.4.4                                                                                                                                                                                                                                                                       korean-lunar-calendar==0.3.1                                                                                                                                                                                                                                                            leather==0.3.4                                                                                                                                                                                                                                                                          locket==1.0.0                                                                                                                                                                                                                                                                           Logbook==1.5.3                                                                                                                                                                                                                                                                          LunarCalendar==0.0.9                                                                                                                                                                                                                                                                    lxml==4.9.2                                                                                                                                                                                                                                                                             MarkupSafe==2.1.2                                                                                                                                                                                                                                                                       mashumaro==3.3.1
matplotlib==3.7.1
minimal-snowplow-tracker==0.0.2
more-itertools==9.1.0
moto==4.1.3
msgpack==1.0.5
multidict==6.0.4
networkx==2.8.8
numpy==1.24.2
openpyxl==3.0.10
opensearch-py==1.1.0
ordered-set==4.1.0
oscrypto==1.3.0
packaging==23.0
pandas==1.3.5
parsedatetime==2.4
partd==1.3.0
pathspec==0.10.3
pg8000==1.22.1
Pillow==9.5.0
pkg_resources==0.0.0
plotly==5.14.1
pluggy==1.0.0
ply==3.11
progressbar2==3.55.0
prophet==1.1
psycopg2-binary==2.9.5
pyarrow==6.0.0
pyathena==2.23.0
pycparser==2.21
pycryptodomex==3.18.0
PyJWT==2.7.0
PyMeeus==0.5.12
PyMySQL==1.0.3
pyOpenSSL==23.1.1
pyparsing==3.0.9
pyrsistent==0.19.3
pytest==7.2.1
pytest-cov==4.0.0
pytest-dependency==0.5.1
pytest-depends==1.0.1
pytest-skip-slow==0.0.5
python-dateutil==2.8.2
python-slugify==8.0.1
python-utils==3.5.2
pytimeparse==1.1.8
pytz==2023.3
PyYAML==6.0
redshift-connector==2.0.910
requests==2.31.0
requests-aws4auth==1.2.2
responses==0.23.1
s3fs==0.4.2
s3transfer==0.6.0
scikit-learn==1.0.2
scipy==1.10.1
scramp==1.4.4
SecretStorage==3.3.3
setuptools-git==1.2
sisense-tools==1.4.1
six==1.16.0
slack-sdk==3.21.0
snowflake-connector-python==3.0.4
sortedcontainers==2.4.0
soupsieve==2.4
sqlparse==0.4.4
stringcase==1.2.0
tabulate==0.9.0
tenacity==8.2.2
text-unidecode==1.3
threadpoolctl==3.1.0
tomli==2.0.1
toolz==0.12.0
tqdm==4.65.0
types-PyYAML==6.0.12.9
typing_extensions==4.5.0
urllib3==1.26.15
Werkzeug==2.2.3
XlsxWriter==3.0.2
xmltodict==0.13.0
yamllint==1.30.0
zipp==3.15.0
```


### What did you do?

```python
pip-audit labels versions of the cryptography package below version 41.0.0 as having a security vulnerability. But the hook pip-compile reads the requirements of snowflake-connector-python[secure-local-storage]==3.0.4 and automatically installs a lower versioned cryptography package, 40.0.2, which is one version previous to cryptography 41.0.0. When pip-compile sees the requirements for snowflake-connector-python[secure-local-storage]==3.0.4 it is specifies that it requires a version lower than cryptography 41.0.0 and causes pip-audit to fail due to the security vulnerability.

requirement=SpecifierRequirement('cryptography<41.0.0,>=3.1.0')

This is a request to bring snowflake-connector-python[secure-local-storage] up to compatibility with the cryptography package 41.0.0 that resolves the security vulnerability.
```


### What did you expect to see?

`pip-audit` labels versions of the `cryptography` package below version 41.0.0 as having a security vulnerability. But the hook `pip-compile` reads the requirements of `snowflake-connector-python[secure-local-storage]==3.0.4` and automatically installs a lower versioned `cryptography` package, 40.0.2, which is one version previous to `cryptography` 41.0.0. When `pip-compile` sees the requirements for `snowflake-connector-python[secure-local-storage]==3.0.4` it is specifies that it requires a version lower than `cryptography 41.0.0` and causes `pip-audit` to fail due to the security vulnerability.

requirement=SpecifierRequirement('cryptography<41.0.0,>=3.1.0')

This is a request to bring `snowflake-connector-python[secure-local-storage]` up to compatibility with the `cryptography` package 41.0.0 that resolves the security vulnerability.

![image](https://github.com/snowflakedb/snowflake-connector-python/assets/118224759/bc76d9f4-ab16-4ebe-864a-f886713a1eb8)

![image](https://github.com/snowflakedb/snowflake-connector-python/assets/118224759/e3f646f6-b6ab-4f6f-b311-071032fed5ea)



### Can you set logging to DEBUG and collect the logs?

```bash
import logging
import os

for logger_name in ('snowflake.connector',):
    logger = logging.getLogger(logger_name)
    logger.setLevel(logging.DEBUG)
    ch = logging.StreamHandler()
    ch.setLevel(logging.DEBUG)
    ch.setFormatter(logging.Formatter('%(asctime)s - %(threadName)s %(filename)s:%(lineno)d - %(funcName)s() - %(levelname)s - %(message)s'))
    logger.addHandler(ch)
```

**Comment by iansantagata** [comment]
Came here to report the same thing and ask for `cryptography~=41.0.0` to be allowed for use with `snowflake-connector-python`.

**Comment by sfc-gh-stan** [comment]
Hi all, `main` has already updated this dependency to be `<42.0.0`, which will be released in `v3.0.5` next week.

**Comment by iansantagata** [comment]
@sfc-gh-stan - do you have a timeline for when this week `v3.0.5` will be released?  Thanks in advance!

**Comment by junnplus** [comment]
@sfc-gh-stan Any progress?

**Comment by bbayles** [comment]
From the `cryptography` documentation's section on [API stability](https://cryptography.io/en/latest/api-stability/#what-does-stable-mean):
> * Public APIs will not be removed or renamed without providing a compatibility alias.
> * The behavior of existing APIs will not change.

Based on this, could the Snowflake Connector For Python team consider changing its policy w.r.t. `cryptography` versions? Specifically, I think it would be better to:
* Define a minimum version that contains the public APIs used by this package
* Do not define a maximum version

This would allow downstream applications to more freely upgrade `cryptography` (an important component for secure applications!) rather than waiting for a corresponding Snowflake Connector For Python release.

The risk is that there is a breaking change in a `cryptography` upgrade (i.e. one that's against their stated policy) that affects applications that blindly upgrade. However, I think this is a risk worth taking:
* Applications shouldn't be blindly upgrading either `cryptography` or `snowflake-connector-python`!
* If there's some fix so important that it requires a breaking change in `cryptography`, it's likely a very serious security issue that ought to break loudly

Hope that's helpful!

**Comment by yhlee-tw** [comment]
according to https://github.com/snowflakedb/snowflake-connector-python/issues/1284#issuecomment-1324299738 and https://github.com/snowflakedb/snowflake-connector-python/pull/1590, shouldn't we bump it to n+3 unless snowflake has pending UserWarning/deprecation to fix?

**Comment by junnplus** [comment]
What are the current issues obstructing the release of the new version? It has been a month and a half since the last release.
@sfc-gh-stan

**Comment by lukehsiao** [comment]
According to https://github.com/snowflakedb/snowflake-connector-python/issues/1640 it should release this week.

**Comment by sfc-gh-achandrasekaran** [comment]
> What are the current issues obstructing the release of the new version? It has been a month and a half since the last release. @sfc-gh-stan

We release the connector around the third week of every month. We have had to push the next release out recently to accommodate a few late breaking customer requests. Once we get them in, we will release a new version of the connector. Thank you for your understanding!

**Comment by lukehsiao** [comment]
@sfc-gh-achandrasekaran when can we expect a release, then? Again, this is a high severity issue: https://pyup.io/v/59062/f17/

If it will be delayed more weeks, it may be prudent to first cut a smaller release with this cryptography patch to unblock those that are waiting for a patch, and then get in the customer requests in a following release.

**Comment by sfc-gh-achandrasekaran** [comment]
@lukehsiao we are targeting Monday at the very latest. If we can get it out before then, we will.

**Comment by lukehsiao** [comment]
@sfc-gh-achandrasekaran is this still planned for today?

**Comment by sfc-gh-achandrasekaran** [comment]
This is now released as part of 3.1.0

**Comment by justinclift** [comment]
Thanks for getting this done @sfc-gh-achandrasekaran. :smile:

**Comment by jitendrachaudhariSiemens** [comment]
[cryptography](https://pypi.org/project/cryptography/)==41.0.7  through 3.1.0 has High/medium vulnerability which has been fixed in 42.0.0 . 
![image](https://github.com/snowflakedb/snowflake-connector-python/assets/156797997/cb239726-4b95-4c1e-a346-1487dd94e33b)

Could you please update this package
